File: read_boundaries.py
"""
read_boundaries.py  —  Python equivalent of read_boundaries.cpp
Reads and prints .rays/boundaries.msgpack
Usage: python read_boundaries.py <project_root>
"""
import sys
from pathlib import Path
import msgpack

def main():
    test_path = sys.argv[1] if len(sys.argv) > 1 else \
        "/Users/samreedhbhuyan/Desktop/Win_C/RAYS/YOLO_model_everything/RAYS_V5/trial_codebases/alphafold"

    msgpack_file = Path(test_path) / ".rays" / "boundaries.msgpack"

    if not msgpack_file.exists():
        print(f"Error: {msgpack_file} does not exist!")
        print("Run 'python rays_generator.py' first to generate it.")
        sys.exit(1)

    with open(msgpack_file, "rb") as f:
        boundary_registry = msgpack.unpack(f, raw=False)

    print("==================================================")
    print(f"Total boundaries found: {len(boundary_registry)}")
    print("==================================================\n")

    for record in boundary_registry:
        print(f"Symbol: {record['symbol_name']}")
        print(f"  Boundary Type: {record['boundary_type']}")
        print(f"  File:          {record['file_path']}")
        print(f"  Line:          {record['line_number']}")
        print(f"  Category:      {record['category']}")
        print(f"  Risk Level:    {record['risk_level']}")
        print("--------------------------------------------------")

    print(f"\nTotal: {len(boundary_registry)} boundaries")

if __name__ == "__main__":
    main()

File: read_boundaries.cpp
#include "rays_generator.cpp"
#include <iostream>
#include <fstream>

int main() {
    std::string test_path = "/Users/samreedhbhuyan/Desktop/Win_C/RAYS/YOLO_model_everything/RAYS_V5/trial_codebases/alphafold";
    
    fs::path msgpack_file = fs::path(test_path) / ".rays" / "boundaries.msgpack";
    
    if (!fs::exists(msgpack_file)) {
        std::cout << "Error: " << msgpack_file << " does not exist!" << std::endl;
        std::cout << "Run './test_rays' first to generate it." << std::endl;
        return 1;
    }
    
    std::ifstream ifs(msgpack_file, std::ios::binary);
    std::string buffer_str((std::istreambuf_iterator<char>(ifs)), std::istreambuf_iterator<char>());
    
    msgpack::object_handle oh = msgpack::unpack(buffer_str.data(), buffer_str.size());
    msgpack::object obj = oh.get();
    
    std::vector<BoundaryRecord> boundary_registry;
    obj.convert(boundary_registry);
    
    std::cout << "==================================================" << std::endl;
    std::cout << "Total boundaries found: " << boundary_registry.size() << std::endl;
    std::cout << "==================================================" << std::endl << std::endl;
    
    for (const auto& record : boundary_registry) {
        std::cout << "Symbol: " << record.symbol_name << std::endl;
        std::cout << "  Boundary Type: " << record.boundary_type << std::endl;
        std::cout << "  File: " << record.file_path << std::endl;
        std::cout << "  Line: " << record.line_number << std::endl;
        std::cout << "  Category: " << record.category << std::endl;
        std::cout << "  Risk Level: " << record.risk_level << std::endl;
        std::cout << "--------------------------------------------------" << std::endl;
    }
    
    std::cout << std::endl << "Total: " << boundary_registry.size() << " boundaries" << std::endl;
    
    return 0;
}

